### PR TITLE
Delay materializing RangeIndex in .reset_index

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Before, a `RangeIndex` would be materialized to a `Column` even if it wasn't used (`drop=True`). Now, it's only materialized if the index will be added as a new column in the DataFrame.

Also caught a validation bug where an `invalid` number of levels would not raise an error

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -955,6 +956,13 @@ def __pos__(self):
     def __abs__(self):
         return abs(self._as_int_index())
 
+    def _columns_for_reset_index(
+        self, levels: tuple | None
+    ) -> Generator[tuple[Any, ColumnBase], None, None]:
+        """Return the columns and column names for .reset_index"""
+        # We need to explicitly materialize the RangeIndex to a column
+        yield "index" if self.name is None else self.name, as_column(self)
```
Comment (mroeschke) [inline python/cudf/cudf/core/index.py:964]: cc @vyasr I hope this is OK to explicitly signal that we need to create a column from the `RangeIndex`

**Review by galipremsagar — COMMENTED** [review]
Nothing jumps out for me while looking at the code but @mroeschke can you look into why there are new failures in the following files:

- tests/generic/test_finalize.py
- tests/frame/methods/test_to_period.py
- tests/frame/methods/test_rank.py

It might be noise but wanted to be sure if these failures are new because of this PR or not.

**Comment by mroeschke** [comment]
> can you look into why there are new failures in the following files:

Did you see these failures locally? Looks like the CI didn't pick up these failures

File: python/cudf/cudf/core/index.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/index.py:944]: Yes I think that is OK. This is specifically for when the index becomes a column in the Frame, right? At which point the materialization is what is explicitly being requested?

**Comment by vyasr** [comment]
@galipremsagar could you comment on where you saw those failures?

**Comment by galipremsagar** [comment]
> @galipremsagar could you comment on where you saw those failures?

They were here: https://github.com/rapidsai/cudf/actions/runs/8823446932/attempts/1#summary-24226360963

But after looking at my PRs, these failures are probably un-related. So ignore my comment.

**Review by vyasr — APPROVED** [review]
Based on Prem's comment above, approving assuming that my understanding in the one question that I raise is correct.

File: python/cudf/cudf/core/index.py (hunk 1 above)
Comment (mroeschke) [reply python/cudf/cudf/core/index.py:979]: Correct, when this is called we want a new column to add to the new DataFrame from this RangeIndex

**Pull request merged**